freshmaker/lightblue.py

```diff
@@ -873,6 +893,67 @@ def filter_out_images_based_on_content_set(self, images, content_sets):
                 ret.append(image)
         return ret
 
+    def _set_container_image_filters(self, request, content_sets,
+                                     rpm_nvrs_names, published):
+        """
+        Sets the additional filters to containerImage request
+
+        :param dict request: request that should be modified
+        :param list content_sets: List of content_sets the image includes RPMs
+            from.
+        :param list rpm_nvrs_names: list names of the binary RPM NVRs to look for
+        :param bool published: whether to limit queries to published
+            repositories(some unpublished repos still will be queried)
+        """
+        if content_sets:
+            request["query"]["$and"][1]["$or"][0]["$and"].append(
+                {
+                    "$or": [{
+                        "field": "content_sets.*",
+                        "op": "=",
+                        "rvalue": r
+                    } for r in content_sets]
+                })
+
+        if rpm_nvrs_names:
+            request["query"]["$and"][1]["$or"][0]["$and"].append(
+                {
+                    "$or": [{
+                        "field": "rpm_manifest.*.rpms.*.name",
+                        "op": "=",
+                        "rvalue": rpm_name
+                    } for rpm_name in rpm_nvrs_names]
+                })
+
+        if published is not None:
+            request["query"]["$and"][1]["$or"][0]["$and"].append(
+                {
+                    "field": "repositories.*.published",
+                    "op": "=",
+                    "rvalue": published
+                })
+            # If the query is for published images, add configurable repos  for
+            # unpublished images(like EUS) too because they shouldn't be ignored
+            if published is True and conf.unpublished_exceptions:
+                for repo in conf.unpublished_exceptions:
+                    request["query"]["$and"][1]["$or"].append(
```

Comment:
You'll also need the other filters that were applied to `request["query"]["$and"][1]["$or"][0]["$and"]`. My suggestion is that you move the `"repositories.*.published"` filter from above to below the `if published is True and conf.unpublished_exceptions:` code block. Then for each repo in `conf.unpublished_exceptions`, you would first deep copy `request["query"]["$and"][1]["$or"][0]["$and"]` and then you would also add the filters for `unpublished_exceptions` to the deep copy. Then you would append the result to `request["query"]["$and"][1]["$or"]`.

Reply:
Are you sure that it will work? Because as I understood if I do as you suggested, I will get at the end wrong filters like:
```
request["query"]["$and"][1]["$or"]:
   [
      {"$or": [....some_rpm_manifests...]},
      {"$or": [....some_content_sets...]},
      {"$and": [...unpublished_repos...]}
   ]
```
but I think I might need to do something like:
```
request["query"]["$and"][1]["$or"]:
   [
      {"$and":
         [
           {"published": False},
           {"registry": "...registry_from_unpublished_exceptions..."}
           {"repository": "...repository_from_unpublished_exceptions..."}
           {"$or": [....some_rpm_manifests...]},
           {"$or": [....some_content_sets...]}
         ]

      },
      {"$and":
         [
           {"published": False},
           {"registry": "...registry#2_from_unpublished_exceptions..."}
           {"repository": "...repository#2_from_unpublished_exceptions..."}
           {"$or": [....some_rpm_manifests...]},
           {"$or": [....some_content_sets...]}
         ]
      },
     ...
 ]
```

Reply:
So I've made a change in the code as I described, you can check if it's ok.